core/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/PutStoredScriptRequest.java

```diff
@@ -92,8 +97,27 @@ public BytesReference script() {
         return script;
     }
 
+    public XContentType xContentType() {
+        return xContentType;
+    }
+
+    /**
+     * Set the script source using bytes.
+     * @deprecated this method is deprecated as it relies on content type detection. Use {@link #script(BytesReference, XContentType)}
+     */
+    @Deprecated
     public PutStoredScriptRequest script(BytesReference source) {
         this.script = source;
+        this.xContentType = Objects.requireNonNull(XContentFactory.xContentType(source));
+        return this;
+    }
+
+    /**
+     * Set the script source and the content type of the bytes.
+     */
+    public PutStoredScriptRequest script(BytesReference source, XContentType xContentType) {
```

Comment:
forget about this comment, scripts are stored in json format, the conversion happens in ScriptMetaData, which is ok I think. we just seem to be sending out via transport whatever format we have, but the master will do the conversion before storing it.